Support setting Firebase private key only at runtime

**Describe the bug**
I'm using a Docker image to deploy my nextjs app however when I run yarn build it seems to expect the private key. This doesn't make sense for my scenario. The Firebase private key will be supplied at runtime as environment variable so it shouldn't be needed during yarn build.

See logs:
```
Step #0: [91m
Step #0: > Build error occurred
Step #0: [0m[91mFirebaseAppError: Service account object must contain a string "private_key" property.
Step #0:     at FirebaseAppError.FirebaseError [as constructor] (/build/node_modules/firebase-admin/lib/utils/error.js:44:28)
Step #0:     at FirebaseAppError.PrefixedFirebaseError [as constructor] (/build/node_modules/firebase-admin/lib/utils/error.js:90:28)
Step #0:     at new FirebaseAppError (/build/node_modules/firebase-admin/lib/utils/error.js:125:28)
Step #0:     at new ServiceAccount (/build/node_modules/firebase-admin/lib/credential/credential-internal.js:134:19)
Step #0:     at new ServiceAccountCredential (/build/node_modules/firebase-admin/lib/credential/credential-internal.js:68:15)
Step #0:     at Object.exports.cert (/build/node_modules/firebase-admin/lib/credential/credential.js:34:54)
Step #0:     at /build/node_modules/next-firebase-auth/build/index.node.js:2:16259
Step #0:     at init (/build/node_modules/next-firebase-auth/build/index.node.js:2:16290)
Step #0:     at initAuth (/build/.next/server/pages/_app.js:2976:66)
Step #0:     at Module.1TCz (/build/.next/server/pages/_app.js:416:36)
Step #0:     at __webpack_require__ (/build/.next/server/pages/_app.js:23:31)
Step #0:     at Object.0 (/build/.next/server/pages/_app.js:116:18)
Step #0:     at __webpack_require__ (/build/.next/server/pages/_app.js:23:31)
Step #0:     at /build/.next/server/pages/_app.js:91:18
Step #0:     at Object.<anonymous> (/build/.next/server/pages/_app.js:94:10)
Step #0:     at Module._compile (node:internal/modules/cjs/loader:1108:14) {
Step #0:   errorInfo: {
Step #0:     code: 'app/invalid-credential',
Step #0:     message: 'Service account object must contain a string "private_key" property.'
Step #0:   },
Step #0:   codePrefix: 'app'
Step #0: }
Step #0: [0m[91merror Command failed with exit code 1.
Step #0: [0minfo Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Step #0: The command '/bin/sh -c export COOKIE_SECRET_CURRENT=$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1) &&     export COOKIE_SECRET_PREVIOUS=$COOKIE_SECRET_CURRENT &&     yarn build &&     rm -rf .next/cache' returned a non-zero code: 1
```


**Version**
Latest, ^0.12.0

Expected result:
Ability to run yarn build without having to set the private key

I followed the example application.

The workaround I tried was to load the firebase admin myself however it looks like next-firebase-auth doesn't support it. I keep getting the following errors as soon as I remove the firebaseAdminConfig while running `yarn dev`:
```
error - ./node_modules/gcs-resumable-upload/node_modules/google-auth-library/build/src/auth/googleauth.js:17:0          
Module not found: Can't resolve 'child_process'                                                                                    
null                                                                                                                               
event - build page: /next/dist/pages/_error                                                                                        
wait  - compiling...
error - ./node_modules/gcs-resumable-upload/node_modules/google-auth-library/build/src/auth/googleauth.js:17:0
Module not found: Can't resolve 'child_process'
null
Error: Cannot find module '/home/samos/workspace/websu-saas/.next/build-manifest.json'
Require stack:
- /home/samos/workspace/websu-saas/node_modules/next/dist/next-server/server/load-components.js
```

--- a/src/__tests__/firebaseAdmin.test.js
+++ b/src/__tests__/firebaseAdmin.test.js
@@ -1,10 +1,12 @@
-import * as admin from 'firebase-admin'
 import { createMockFirebaseUserAdminSDK } from 'src/testHelpers/authUserInputs'
 import createMockFetchResponse from 'src/testHelpers/createMockFetchResponse'
 import createAuthUser from 'src/createAuthUser'
 import { setConfig } from 'src/config'
 import createMockConfig from 'src/testHelpers/createMockConfig'
+import getFirebaseAdminApp from 'src/initFirebaseAdminSDK'
 
+// We're not mocking initFirebaseAdminSDK.js, instead just mocking
+// the underyling Firebase admin app.
 jest.mock('firebase-admin')
 
 beforeEach(() => {
@@ -33,6 +35,7 @@ describe('verifyIdToken', () => {
       firebaseUserAdminSDK: mockFirebaseUser,
       token: 'some-token',
     })
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     const response = await verifyIdToken('some-token')
     expect(response).toEqual({
@@ -46,6 +49,7 @@ describe('verifyIdToken', () => {
   it('returns an AuthUser with the same token when the token has not expired', async () => {
     const { verifyIdToken } = require('src/firebaseAdmin')
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     const AuthUser = await verifyIdToken('some-token')
     const token = await AuthUser.getIdToken()
@@ -73,6 +77,7 @@ describe('verifyIdToken', () => {
     )
     expiredTokenErr.code = 'auth/id-token-expired'
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockImplementation(async (token) => {
       if (token === 'some-token') {
         throw expiredTokenErr
@@ -104,6 +109,7 @@ describe('verifyIdToken', () => {
     )
     expiredTokenErr.code = 'auth/id-token-expired'
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockImplementation(async (token) => {
       if (token === 'some-token') {
         throw expiredTokenErr
@@ -126,6 +132,7 @@ describe('verifyIdToken', () => {
     )
     expiredTokenErr.code = 'auth/id-token-expired'
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockImplementation(async (token) => {
       if (token === 'some-token') {
         throw expiredTokenErr
@@ -160,6 +167,7 @@ describe('verifyIdToken', () => {
     )
     expiredTokenErr.code = 'auth/id-token-expired'
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockImplementation(async (token) => {
       if (token === 'some-token') {
         throw expiredTokenErr
@@ -194,6 +202,7 @@ describe('verifyIdToken', () => {
     // Mock that the original token is expired but a new token works.
     const otherErr = new Error('The Firebase ID token has been revoked.')
     otherErr.code = 'auth/id-token-revoked' // a different error
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockImplementation(async () => {
       throw otherErr
     })
@@ -211,6 +220,7 @@ describe('verifyIdToken', () => {
       'The provided Firebase ID token is expired.'
     )
     expiredTokenErr.code = 'auth/id-token-expired'
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockImplementation(async () => {
       throw expiredTokenErr
     })
@@ -224,6 +234,7 @@ describe('getCustomIdAndRefreshTokens', () => {
   it("passes the Firebase user's ID (from verifyIdToken) to createCustomToken", async () => {
     const { getCustomIdAndRefreshTokens } = require('src/firebaseAdmin')
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     admin.auth().createCustomToken.mockResolvedValue('my-custom-token')
     await getCustomIdAndRefreshTokens('some-token')
@@ -247,6 +258,7 @@ describe('getCustomIdAndRefreshTokens', () => {
     })
 
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     admin.auth().createCustomToken.mockResolvedValue('my-custom-token')
     await getCustomIdAndRefreshTokens('some-token')
@@ -260,6 +272,7 @@ describe('getCustomIdAndRefreshTokens', () => {
   it('uses the expected fetch options when calling to get a custom token', async () => {
     const { getCustomIdAndRefreshTokens } = require('src/firebaseAdmin')
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     admin.auth().createCustomToken.mockResolvedValue('my-custom-token')
     await getCustomIdAndRefreshTokens('some-token')
@@ -287,6 +300,7 @@ describe('getCustomIdAndRefreshTokens', () => {
     })
 
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     admin.auth().createCustomToken.mockResolvedValue('my-custom-token')
     const response = await getCustomIdAndRefreshTokens('some-token')
@@ -314,6 +328,7 @@ describe('getCustomIdAndRefreshTokens', () => {
     const expectedAuthUser = createAuthUser({
       firebaseUserAdminSDK: mockFirebaseUser,
     })
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     admin.auth().createCustomToken.mockResolvedValue('my-custom-token')
     const response = await getCustomIdAndRefreshTokens('some-token')
@@ -341,6 +356,7 @@ describe('getCustomIdAndRefreshTokens', () => {
     })
 
     const mockFirebaseUser = createMockFirebaseUserAdminSDK()
+    const admin = getFirebaseAdminApp()
     admin.auth().verifyIdToken.mockResolvedValue(mockFirebaseUser)
     admin.auth().createCustomToken.mockResolvedValue('my-custom-token')
     await expect(getCustomIdAndRefreshTokens('some-token')).rejects.toEqual(

--- a/src/__tests__/index.server.test.js
+++ b/src/__tests__/index.server.test.js
@@ -23,11 +23,13 @@ describe('index.server.js: init', () => {
     expect(indexServer.init).toEqual(expect.any(Function))
   })
 
-  it('calls initFirebaseAdminSDK', () => {
+  // We only initialize the Firebase admin SDK as it's needed. See:
+  // https://github.com/gladly-team/next-firebase-auth/issues/70
+  it('does not call initFirebaseAdminSDK', () => {
     expect.assertions(1)
     const indexServer = require('src/index.server').default
     indexServer.init({ fake: 'config' })
-    expect(initFirebaseAdminSDK).toHaveBeenCalled()
+    expect(initFirebaseAdminSDK).not.toHaveBeenCalled()
   })
 
   it('calls index.js (client) init', () => {

--- a/src/__tests__/initFirebaseAdminSDK.test.js
+++ b/src/__tests__/initFirebaseAdminSDK.test.js
@@ -36,6 +36,13 @@ describe('initFirebaseAdminSDK', () => {
     })
   })
 
+  it('returns the admin app', () => {
+    expect.assertions(1)
+    const initFirebaseAdminSDK = require('src/initFirebaseAdminSDK').default
+    const response = initFirebaseAdminSDK()
+    expect(response).toEqual(admin)
+  })
+
   it('does not call admin.initializeApp if Firebase already has an initialized app', () => {
     expect.assertions(1)
     admin.apps = [{ some: 'app' }]

--- a/src/firebaseAdmin.js
+++ b/src/firebaseAdmin.js
@@ -1,4 +1,4 @@
-import * as admin from 'firebase-admin'
+import getFirebaseAdminApp from 'src/initFirebaseAdminSDK'
 import createAuthUser from 'src/createAuthUser'
 import { getConfig } from 'src/config'
 
@@ -48,6 +48,7 @@ const refreshExpiredIdToken = async (refreshToken) => {
 export const verifyIdToken = async (token, refreshToken = null) => {
   let firebaseUser
   let newToken = token
+  const admin = getFirebaseAdminApp()
   try {
     firebaseUser = await admin.auth().verifyIdToken(token)
   } catch (e) {
@@ -81,6 +82,7 @@ export const verifyIdToken = async (token, refreshToken = null) => {
  */
 export const getCustomIdAndRefreshTokens = async (token) => {
   const AuthUser = await verifyIdToken(token)
+  const admin = getFirebaseAdminApp()
 
   // It's important that we pass the same user ID here, otherwise
   // Firebase will create a new user.

--- a/src/index.server.js
+++ b/src/index.server.js
@@ -5,15 +5,12 @@ import index from 'src/index'
 import setAuthCookies from 'src/setAuthCookies'
 import unsetAuthCookies from 'src/unsetAuthCookies'
 import withAuthUserTokenSSRModule from 'src/withAuthUserTokenSSR'
-import initFirebaseAdminSDK from 'src/initFirebaseAdminSDK'
 import { verifyIdToken } from 'src/firebaseAdmin'
 
 const initServer = (config) => {
   const clientInit = index.init(config)
-
-  // On the server only, initialize the Firebase admin SDK.
-  initFirebaseAdminSDK()
-
+  // We only initialize the Firebase admin SDK as it's needed. See:
+  // https://github.com/gladly-team/next-firebase-auth/issues/70
   return clientInit
 }
 

--- a/src/initFirebaseAdminSDK.js
+++ b/src/initFirebaseAdminSDK.js
@@ -16,6 +16,7 @@ const initFirebaseAdminSDK = () => {
       }),
     })
   }
+  return admin
 }
 
 export default initFirebaseAdminSDK

